src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -55,69 +60,190 @@ protected override string CreateRequestId()
                 host = authority;
             }
 
-            // TODO: OPTIONS * requests?
-            // To ensure that the HTTP / 1.1 request line can be reproduced
-            // accurately, this pseudo - header field MUST be omitted when
-            // translating from an HTTP/ 1.1 request that has a request target in
-            // origin or asterisk form(see[RFC7230], Section 5.3).
-            // https://tools.ietf.org/html/rfc7230#section-5.3
-
-            if (host.Count <= 0)
+            // https://tools.ietf.org/html/rfc7230#section-5.4
+            // A server MUST respond with a 400 (Bad Request) status code to any
+            // HTTP/1.1 request message that lacks a Host header field and to any
+            // request message that contains more than one Host header field or a
+            // Host header field with an invalid field-value.
+            hostText = host.ToString();
+            if (host.Count > 1 || !HttpUtilities.IsHostHeaderValid(hostText))
             {
-                BadHttpRequestException.Throw(RequestRejectionReason.MissingHostHeader);
+                // RST replaces 400
+                ResetAndAbort(new ConnectionAbortedException(CoreStrings.FormatBadRequest_InvalidHostHeader_Detail(hostText)), Http2ErrorCode.PROTOCOL_ERROR);
+                return false;
             }
-            else if (host.Count > 1)
+
+            return true;
+        }
+
+        private bool TryValidatePath(ReadOnlySpan<char> pathSegment)
+        {
+            // Must start with a leading slash
+            if (pathSegment.Length == 0 || pathSegment[0] != '/')
             {
-                BadHttpRequestException.Throw(RequestRejectionReason.MultipleHostHeaders);
+                ResetAndAbort(new ConnectionAbortedException(CoreStrings.FormatHttp2StreamErrorPathInvalid(RawTarget)), Http2ErrorCode.PROTOCOL_ERROR);
+                return false;
             }
 
-            var hostText = host.ToString();
-            HttpUtilities.ValidateHostHeader(hostText);
+            var pathEncoded = pathSegment.IndexOf('%') >= 0;
 
-            endConnection = false;
-            return true;
+            // Compare with Http1Connection.OnOriginFormTarget
+
+            // URIs are always encoded/escaped to ASCII https://tools.ietf.org/html/rfc3986#page-11
+            // Multibyte Internationalized Resource Identifiers (IRIs) are first converted to utf8;
+            // then encoded/escaped to ASCII  https://www.ietf.org/rfc/rfc3987.txt "Mapping of IRIs to URIs"
+
+            try
+            {
+                // The decoder operates only on raw bytes
+                var pathBuffer = new byte[pathSegment.Length].AsSpan();
+                for (int i = 0; i < pathSegment.Length; i++)
+                {
+                    var ch = pathSegment[i];
+                    // The header parser should already be checking this
+                    Debug.Assert(20 < ch && ch < 127);
```

Comment:
This looks odd. If the header parser is checking this then why bother with the debug.assert? if we want to double check, then I think we'll need to follow the RFC https://tools.ietf.org/html/rfc3986#section-3.3 and reject gen-delims for example (for example `[` and `]`). If we want to do a quick check I think you want to assert `32 < ch && ch < 127` 20-32 in dec are still control chars and space.

Reply:
The point of a Debug.Assert is that it should be impossible for it to fail. I don't mind double checking at all. What we shouldn't do is add a Debug.Assert that could fail because previous validation was less strict than the assert.